core/vm/evmc.go

```diff
@@ -0,0 +1,307 @@
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package vm
+
+import (
+	"fmt"
+	"math/big"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/ethereum/evmc/bindings/go/evmc"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/params"
+)
+
+type EVMC struct {
+	instance *evmc.Instance
+	env      *EVM
+	readOnly bool // TODO: The readOnly flag should not be here.
+}
+
+var (
+	createMu     sync.Mutex
+	evmcInstance *evmc.Instance
+)
+
+func createVM() *evmc.Instance {
+	createMu.Lock()
+	defer createMu.Unlock()
+
+	if evmcInstance == nil {
+		vmPath := os.Getenv("EVMC_PATH")
+		if len(vmPath) == 0 {
+			panic("EVMC VM path not provided, set EVMC_PATH environment variable")
+		}
+
+		var err error
+		evmcInstance, err = evmc.Load(vmPath)
+		if err != nil {
+			panic(err.Error())
+		}
+		log.Info("EVMC VM loaded", "name", evmcInstance.Name(), "version", evmcInstance.Version(), "path", vmPath)
+
+		for _, option := range strings.Split(os.Getenv("EVMC_OPTIONS"), " ") {
+			if idx := strings.Index(option, "="); idx >= 0 {
+				name := option[:idx]
+				value := option[idx+1:]
+				err := evmcInstance.SetOption(name, value)
+				if err == nil {
+					log.Info("EVMC VM option set", "name", name, "value", value)
+				} else {
+					log.Warn("EVMC VM option setting failed", "name", name, "error", err)
+				}
+			}
+		}
+	}
+	return evmcInstance
+}
+
+func NewEVMC(env *EVM) *EVMC {
+	return &EVMC{createVM(), env, false}
+}
+
+// Implements evmc.HostContext interface.
+type HostContext struct {
+	env      *EVM
+	contract *Contract
+}
+
+func (host *HostContext) AccountExists(addr common.Address) bool {
+	env := host.env
+	eip158 := env.ChainConfig().IsEIP158(env.BlockNumber)
+	if eip158 {
+		if !env.StateDB.Empty(addr) {
+			return true
+		}
+	} else if env.StateDB.Exist(addr) {
+		return true
+	}
+	return false
+}
+
+func (host *HostContext) GetStorage(addr common.Address, key common.Hash) common.Hash {
+	env := host.env
+	return env.StateDB.GetState(addr, key)
+}
+
+func (host *HostContext) SetStorage(addr common.Address, key common.Hash, value common.Hash) (status evmc.StorageStatus) {
+	env := host.env
+
+	oldValue := env.StateDB.GetState(addr, key)
+	if oldValue == value {
+		return evmc.StorageUnchanged
+	}
+
+	env.StateDB.SetState(addr, key, value)
+
+	zero := common.Hash{}
+	status = evmc.StorageModified
+	if oldValue == zero {
+		return evmc.StorageAdded
+	} else if value == zero {
+		env.StateDB.AddRefund(params.SstoreRefundGas)
+		return evmc.StorageDeleted
+	}
+	return evmc.StorageModified
+}
+
+func (host *HostContext) GetBalance(addr common.Address) common.Hash {
```

Comment:
Rename to getExternalBalance

Reply:
It matches the EVMC naming: https://github.com/ethereum/evmc/blob/master/include/evmc/evmc.h#L485 and I want to match the names exactly.

I don't believe the "external" is needed here because it is implicit for all Host methods (all of them takes the address as an argument).